### PR TITLE
Adding Usage Examples to Splitters

### LINKED ISSUE #3230: Add usage example documentation for splitters
Many splitters in `dc.splitters` lack example documentation. We should add documentation to those to make splitters more useful.

### PR DESCRIPTION
**Pull request by GreatRSingh**
# Pull Request Template

## Description

Fix #3230

This PR aims to add example documentation for Splitters.
- [x] RandomSplitter - Already Done
- [x] RandomGroupSplitter
- [x] RandomStratifiedSplitter
- [x] SingletaskStratifiedSplitter - Already Done
- [x] IndexSplitter
- [x] SpecifiedSplitter
- [x] MolecularWeightSplitter
- [x] MaxMinSplitter
- [x] ButinaSplitter
- [x] FingerprintSplitter
- [x] ScaffoldSplitter - Already Done

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by GreatRSingh** [comment]
@gusty1g sir, check the example for MolecularWeightSplitter.

File: deepchem/splits/splitters.py [hunk 1]
```diff
@@ -763,6 +763,18 @@ class IndexSplitter(Splitter):
...
+    Examples
+    --------
+    >>> n_samples = 100
+    >>> n_features = 10
+    >>> n_tasks = 10
+    >>> X = np.random.rand(n_samples, n_features)
+    >>> y = np.random.rand(n_samples, n_tasks)
+    >>> splitter = dc.splits.IndexSplitter()
+    >>> dataset = dc.data.NumpyDataset(X, y)
+    >>> train_dataset, valid_dataset , test_dataset = splitter.split(dataset)
```
Comment (gusty1g) [inline deepchem/splits/splitters.py:776]: A `splitter.split` method will return indices for split and not datasets. Instead, if we use `train_dataset, test_dataset = splitter.train_test_split(dataset)`, it will be more correct.

Also, the example should demonstrate how the splitter works. I would rewrite it something like this:
```
n_samples = 5
n_features = 2
X = np.random.rand(n_samples, n_features)
y = np.random.rand(n_samples)
splitter = dc.splits.IndexSplitter()
dataset = dc.data.NumpyDataset(X, y)

>>> train_dataset, test_dataset = splitter.train_test_split(dataset)

>>> print(train_dataset.ids)
[0, 1, 2]
>>> print (test_dataset.ids)
[3, 4]
```
The above example demonstrates how the splitter splits by index.
```

File: deepchem/splits/splitters.py [hunk 2]
```diff
@@ -815,6 +827,18 @@ class SpecifiedSplitter(Splitter):
...
+    Examples
+    --------
+    >>> n_samples = 10
+    >>> n_features = 3
+    >>> n_tasks = 1
+    >>> X = np.random.rand(n_samples, n_features)
+    >>> y = np.random.rand(n_samples, n_tasks)
+    >>> splitter = dc.splits.SpecifiedSplitter(valid_indices=[1,3,5], test_indices=[0,2,7,9])
+    >>> dataset = dc.data.NumpyDataset(X, y)
+    >>> train_dataset, valid_dataset , test_dataset = splitter.split(dataset)
```
Comment (gusty1g) [inline deepchem/splits/splitters.py:840]: Same as comment for IndexSplitter.

File: deepchem/splits/splitters.py [hunk 3]
```diff
@@ -893,6 +917,23 @@ class MolecularWeightSplitter(Splitter):
...
     ----
     This class requires RDKit to be installed.
+
+    Examples
+    --------
+
+    >>> import deepchem as dc
+    >>> import numpy as np
+    >>> # creation of demo data set with some smiles strings
+    >>> data_test= ["CC(C)Cl" , "CCC(C)CO" ,  "CCCCCCCO" , "CCCCCCCC(=O)OC" , "c3ccc2nc1ccccc1cc2c3" , "Nc2cccc3nc1ccccc1cc23" , "C1CCCCCC1" ]
```
Comment (gusty1g) [inline deepchem/splits/splitters.py:927]: We can use a simple example of 5 data points, just carbon atoms `['C', 'CC', 'CCC', 'CCCC', 'CCCCC]`

File: deepchem/splits/splitters.py [hunk 4]
```diff
@@ -893,6 +917,23 @@ class MolecularWeightSplitter(Splitter):
...
+    >>> Xs = np.zeros(len(data_test))
+    >>> Ys = np.ones(len(data_test))
+    >>> # creation of a deepchem dataset with the smile codes in the ids field
+    >>> dataset = dc.data.DiskDataset.from_numpy(X=Xs,y=Ys,w=np.zeros(len(data_test)),ids=data_test)
```
Comment (gusty1g) [inline deepchem/splits/splitters.py:931]: y and w are optional arguments. We can make it simple as `from_numpy(X=data_test, ids=data_test)`. Maybe we can also rename `data_test` as `smiles`.

File: deepchem/splits/splitters.py (hunk 1 above)
Comment (GreatRSingh) [reply deepchem/splits/splitters.py:776]: Thank u for your correction, i have fixed that.

Comment (GreatRSingh) [reply deepchem/splits/splitters.py]: I mistook split function for something else.
I fixed that in the latest commit.

File: deepchem/splits/splitters.py (hunk 3 above)
Comment (GreatRSingh) [reply deepchem/splits/splitters.py]: ok

**Comment by GreatRSingh** [comment]
@gusty1g can you suggest example for RandomGroupSplitter. I am having difficulty in making example for it.

File: deepchem/splits/splitters.py (hunk 4 above)
Comment (GreatRSingh) [reply deepchem/splits/splitters.py]: fixed

**Comment by gusty1g** [comment]
> @gusty1g can you suggest example for RandomGroupSplitter. I am having difficulty in making example for it. In the group parameter.

```python3
import deepchem as dc
import numpy as np

groups = [0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3]
splitter = dc.splits.RandomGroupSplitter(groups=groups)

dataset = dc.data.NumpyDataset(X=np.arange(12))   # 12 elements
train, test = splitter.train_test_split(dataset, frac_train=0.75, seed=0)

print (train.ids)  # array([6, 7, 8, 9, 10, 11, 3, 4, 5], dtype=object) 
```
In above example, note that the values `6, 7, 8` belong to a group (2), `9, 10, 11` belong to another group (3) and 3 4 5 belong to group 1. The splitter preserves the group (i.e a group goes either into train or test) while splitting the dataset.

**Review by gusty1g — APPROVED** [review]
